sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,240 @@
+
+These parameters are defined in plugin's header file.
+
+4. ``GradientLayer::onFootprintChanged()`` just resets ``need_recalculation_`` value.
+
+5. ``GradientLayer::reset()`` method is dummy: it is not used in this example plugin. It remaining there since pure virtual function ``reset()`` in parent ``Layer`` class required to be overriden.
+
+2- Export and make GradientLayer plugin
+---------------------------------------
+
+The written plugin will be loaded in runtime as it's basic parent class and then will be called by plugin handling modules (for costmap2d by ``LayeredCostmap``). This is not a well-known dynamic library loading mechanism. Instead of it the pluginlib opens a given plugin in run-time and provides methods from exported classes to be callable. The mechanism of class exporting tells pluginlib which basic class should be used during these calls. This allows to extend an application by plugins without knowing application source code or recompiling it.
+
+In our example the ``nav2_gradient_costmap_plugin::GradientLayer`` plugin's class should be dynamically loaded as a ``nav2_costmap_2d::Layer`` basic class. For this the plugin should be registered as follows:
+
+1. Plugin's class should be registered with a basic type of loaded class. For this there is a special macro ``PLUGINLIB_EXPORT_CLASS`` should be added to any sorce-file composing the plugin library:
+
+.. code-block:: text
+
+  #include "pluginlib/class_list_macros.hpp"
+  PLUGINLIB_EXPORT_CLASS(nav2_gradient_costmap_plugin::GradientLayer, nav2_costmap_2d::Layer)
+
+This part is usually placed at the end of cpp-file where the plugin class was written (in our example ``gradient_layer.cpp``). 
+
+2. Plugin's inormation should be stored to plugin description file. This is done by using separate XML (in our example ``gradient_plugins.xml``) in the plugin's package. This file contains information about:
+
+ - ``path``: Path and name of library where plugin is placed.
+ - ``name``: Plugin type referenced in ``plugin_types`` parameter (see next section for more details). It could be whatever you want.
+ - ``type``: Plugin class with namespace taken from the source code.
+ - ``basic_class_type``: Basic parent class from which plugin class was derived.
+ - ``description``: Plugin description in a text form.
+
+.. code-block:: xml
+
+  <library path="nav2_gradient_costmap_plugin_core">
+    <class name="nav2_gradient_costmap_plugin/GradientLayer" type="nav2_gradient_costmap_plugin::GradientLayer" base_class_type="nav2_costmap_2d::Layer">
+      <description>This is an example plugin which puts repeating costs gradients to costmap</description>
+    </class>
+  </library>
+
+The export of plugin is performed by including ``pluginlib_export_plugin_description_file()`` cmake-function into ``CMakeLists.txt``. This function installs plugin description file into ``share`` directory and sets ament indexes for plugin description XML to be discoverable as a plugin of selected type:
```

Comment:
Is that true that it also installs the xml? I did not know that.

Reply:
Yes, I've found it in `pluginlib/cmake/pluginlib_export_plugin_description_file.cmake`:
```
macro(pluginlib_export_plugin_description_file plugin_category relative_filename)
  ...
  install(FILES ${relative_filename} DESTINATION share/${PROJECT_NAME}/${relative_dir})
```

Reply:
Awesome, I can remove other install calls for the XML then.